File: ios/PTimerKit/Sources/PTimerKit/AlternateReciprocityModels.swift
import Foundation

/// Registry of alternate reciprocity profiles/models that live OUTSIDE
/// the launch preset catalog (which enforces exactly one official
/// profile per film). These are selectable through the model selector
/// (the main-screen segmented control, mirrored in Reciprocity Details)
/// via `selectedProfileOverride`, never as duplicate top-level film
/// rows (PTIMER-159).
///
/// Films that currently expose an alternate model:
/// - Kodak Portra 400 — an unofficial practical approximation
///   (lower authority; defined in `UnofficialPracticalProfiles`).
/// - Fomapan 100 Classic — two non-default alternates alongside the
///   default official FOMA table log-log model: the Ohzart community
///   practical table (a table-derived model that reproduces the
///   community anchors exactly) and an APP-DERIVED formula fitted to the
///   official FOMA table. Both are clearly labelled and never presented
///   as the official source/model.
public enum AlternateReciprocityModels {

    /// Alternate models for a film stock, in display order. Empty when
    /// the film has only its single catalog profile.
    public static func alternates(forFilmID filmID: String) -> [ReciprocityProfile] {
        switch filmID {
        case "kodak-portra-400":
            return [UnofficialPracticalProfiles.kodakPortra400UnofficialPractical]
        case "foma-fomapan-100":
            // Display order after the default official table:
            // Ohzart community table, then the app-derived formula.
            return [fomapan100OhzartCommunityTable, fomapan100AppDerivedFormula]
        case "kodak-tri-x-400":
            // Display order after the default official graph/table:
            // the published-rows-only official table, then the
            // app-derived formula fitted to those three rows.
            return [triX400OfficialTable, triX400AppDerivedFormula]
        default:
            return []
        }
    }

    /// `true` for explicitly app-derived alternate models (a formula the
    /// app fitted to a manufacturer table). The "App-derived comparison"
    /// section is intentionally limited to these enrolled models, so it
    /// never leaks onto non-app-derived profiles that merely carry
    /// source anchors — source-backed table/graph models (e.g. Tri-X's
    /// Official Kodak graph/table) or converted-formula profiles (e.g.
    /// Provia). Any future app-derived alternate must be enrolled here
    /// to surface its comparison (PTIMER-159).
    public static func isAppDerivedModel(id: String) -> Bool {
        id == fomapan100AppDerivedFormula.id
            || id == triX400AppDerivedFormula.id
    }

    /// Resolves an alternate profile by its id (used by session restore
    /// to reconstruct a persisted profile override).
    public static func profile(withID profileID: String) -> ReciprocityProfile? {
        let all = [
            UnofficialPracticalProfiles.kodakPortra400UnofficialPractical,
            fomapan100OhzartCommunityTable,
            fomapan100AppDerivedFormula,
            triX400OfficialTable,
            triX400AppDerivedFormula,
        ]
        return all.first { $0.id == profileID }
    }

    /// Ohzart community practical table for Fomapan 100 — an
    /// unofficial, test-based community source (not FOMA-published).
    /// Modelled as a TABLE, not a fitted formula: the published anchors
    /// (1s→1.9s … 60s→795s) are reproduced exactly by the same log-log
    /// interpolation the official FOMA table uses, so every Ohzart row
    /// comes back without fitting error. It shares Official FOMA's
    /// 0.5 s no-correction boundary; its published range ends at 60 s,
    /// past which the model extrapolates the last segment and presents
    /// the value as beyond source range. Kept non-default and labelled
    /// "Ohzart"; its source rows stay separate from the official FOMA
    /// table and never read as manufacturer data.
    public static let fomapan100OhzartCommunityTable = ReciprocityProfile(
        id: "foma-fomapan-100-ohzart-community-table",
        name: "Ohzart community table",
        source: ReciprocitySourceProvenance(
            kind: .thirdPartyPublication,
            authority: .unofficial,
            confidence: .medium,
            publisher: "Ohzart",
            title: "Reciprocity practical table",
            citation: "https://ohzart1.tistory.com/78"
        ),
        rules: [
            .tableInterpolation(TableInterpolationReciprocityRule(
                anchors: [
                    TableAnchor(meteredSeconds: 1, correctedSeconds: 1.9),
                    TableAnchor(meteredSeconds: 2, correctedSeconds: 5),
                    TableAnchor(meteredSeconds: 4, correctedSeconds: 13),
                    TableAnchor(meteredSeconds: 8, correctedSeconds: 35),
                    TableAnchor(meteredSeconds: 15, correctedSeconds: 90),
                    TableAnchor(meteredSeconds: 30, correctedSeconds: 265),
                    TableAnchor(meteredSeconds: 60, correctedSeconds: 795),
                ],
                notes: [
                    "Ohzart community practical table for Fomapan 100, reproduced by log-log interpolation between the published anchors. Practical / community guidance, not FOMA-published data.",
                ],
                noCorrectionThroughSeconds: 0.5,
                sourceRangeThroughSeconds: 60
            )),
        ],
        notes: [
            "Unofficial practical community table (Ohzart). Not FOMA-published data.",
        ],
        sourceEvidence: ohzartCommunityAnchorEvidence,
        modelBasis: ReciprocityProfileModelBasis(
            sourceModel: .practicalCommunityGuidance,
            calculationModel: .tableLogLogInterpolation
        ),
        selectorLabel: "Ohzart"
    )

    /// The Ohzart anchor rows preserved as source evidence so the
    /// "Source reference" section and graph markers show the community
    /// table values (kept distinct from the official FOMA anchors).
    /// Ohzart publishes corrected times directly, so each row carries
    /// only a corrected-time mapping — no multiplier column.
    private static let ohzartCommunityAnchorEvidence: [ReciprocitySourceEvidenceRow] = [
        ohzartAnchorEvidence(metered: 1, corrected: 1.9),
        ohzartAnchorEvidence(metered: 2, corrected: 5),
        ohzartAnchorEvidence(metered: 4, corrected: 13),
        ohzartAnchorEvidence(metered: 8, corrected: 35),
        ohzartAnchorEvidence(metered: 15, corrected: 90),
        ohzartAnchorEvidence(metered: 30, corrected: 265),
        ohzartAnchorEvidence(metered: 60, corrected: 795),
    ]

    private static func ohzartAnchorEvidence(
        metered: Double,
        corrected: Double
    ) -> ReciprocitySourceEvidenceRow {
        ReciprocitySourceEvidenceRow(
            meteredExposure: .exactSeconds(metered),
            adjustments: [
                .exposure(.correctedTime(CorrectedTimeMapping(meteredSeconds: metered, correctedSeconds: corrected))),
            ]
        )
    }

    /// App-derived guarded formula for Fomapan 100 — the retired free
    /// log-log fit `Tc = 2.2457 × Tm^1.4515`. Source is the official
    /// FOMA table (so it cites FOMA), but the calculation is the app's
    /// own fitted formula, not manufacturer-published guidance. Kept
    /// non-default and clearly labelled; selecting it surfaces the
    /// app-derived comparison against the official anchors.
    public static let fomapan100AppDerivedFormula = ReciprocityProfile(
        id: "foma-fomapan-100-app-formula",
        name: "App-derived formula",
        source: ReciprocitySourceProvenance(
            kind: .manufacturerPublished,
            authority: .official,
            confidence: .high,
            publisher: "FOMA BOHEMIA",
            title: "FOMAPAN 100 CLASSIC — Technical sheet",
            citation: "Foma technical sheet"
        ),
        rules: [
            .formula(FormulaReciprocityRule(
                formula: ReciprocityFormula(
                    formulaFamily: .modifiedSchwarzschild,
                    coefficientSeconds: 2.2457,
                    exponent: 1.4515,
                    noCorrectionThroughSeconds: 0.5,
                    sourceRangeThroughSeconds: 100.0
                ),
                notes: [
                    "App-derived: Tc = 2.2457 × Tm^1.4515, a free log-log fit through FOMA's published 1/10/100 sec anchors. Not manufacturer-published guidance; the official table model is the default.",
                ]
            )),
        ],
        sourceEvidence: fomapanOfficialAnchorEvidence,
        modelBasis: ReciprocityProfileModelBasis(
            sourceModel: .manufacturerTable,
            calculationModel: .guardedFormula
        )
    )

    /// The official FOMA anchor rows, preserved as source evidence so a
    /// formula-model comparison can be shown against the published data.
    private static let fomapanOfficialAnchorEvidence: [ReciprocitySourceEvidenceRow] = [
        anchorEvidence(metered: 1, multiplier: 2, corrected: 2, note: "1 sec → ×2 (corrected 2 sec)."),
        anchorEvidence(metered: 10, multiplier: 8, corrected: 80, note: "10 sec → ×8 (corrected 80 sec)."),
        anchorEvidence(metered: 100, multiplier: 16, corrected: 1600, note: "100 sec → ×16 (corrected 1600 sec)."),
    ]

    private static func anchorEvidence(
        metered: Double,
        multiplier: Double,
        corrected: Double,
        note: String
    ) -> ReciprocitySourceEvidenceRow {
        ReciprocitySourceEvidenceRow(
            meteredExposure: .exactSeconds(metered),
            adjustments: [
                .exposure(.multiplier(MultiplierAdjustment(factor: multiplier))),
                .exposure(.correctedTime(CorrectedTimeMapping(meteredSeconds: metered, correctedSeconds: corrected))),
            ],
            notes: [note]
        )
    }

    // MARK: - Kodak Tri-X 400 alternate models (PTIMER-168)

    /// Official Kodak table model for Tri-X 400 using ONLY the three
    /// published E-31 table rows (1/10/100 sec). The no-correction band
    /// ends at 1/10 sec; corrected exposure is log-log interpolated
    /// between the published anchors. Offered as a non-default alternate
    /// to the graph/table model — never presented as a formula.
    public static let triX400OfficialTable = ReciprocityProfile(
        id: "kodak-tri-x-official-table",
        name: "Official Kodak table",
        source: ReciprocitySourceProvenance(
            kind: .manufacturerPublished,
            authority: .official,
            confidence: .high,
            publisher: "Kodak",
            title: "KODAK PROFESSIONAL TRI-X 400 Film — Technical Data",
            citation: "Publication F-4017"
        ),
        rules: [
            .tableInterpolation(TableInterpolationReciprocityRule(
                anchors: [
                    TableAnchor(meteredSeconds: 1, correctedSeconds: 2),
                    TableAnchor(meteredSeconds: 10, correctedSeconds: 50),
                    TableAnchor(meteredSeconds: 100, correctedSeconds: 1200),
                ],
                notes: [
                    "Published Kodak E-31 table rows only (1 sec → 2 sec, 10 sec → 50 sec, 100 sec → 1200 sec). No adjustment through 1/10 sec; log-log interpolation between the published anchors. Inputs above 100 sec are flagged beyond the published source range.",
                ],
                noCorrectionThroughSeconds: 0.1,
                sourceRangeThroughSeconds: 100
            )),
        ],
        sourceEvidence: triX400OfficialAnchorEvidence,
        modelBasis: ReciprocityProfileModelBasis(
            sourceModel: .manufacturerTable,
            calculationModel: .tableLogLogInterpolation
        ),
        selectorLabel: "Table"
    )

    /// App-derived formula for Tri-X 400 — `Tc = 2 × Tm^1.3891`, fitted to
    /// Kodak's three published table rows. Source is the Kodak table (so
    /// it cites Kodak), but the calculation is the app's own fit, NOT a
    /// Kodak-published formula. Kept non-default and clearly labelled; it
    /// is enrolled in `isAppDerivedModel` so it surfaces the app-derived
    /// comparison rather than reading as manufacturer guidance.
    public static let triX400AppDerivedFormula = ReciprocityProfile(
        id: "kodak-tri-x-app-formula",
        name: "App formula",
        source: ReciprocitySourceProvenance(
            kind: .manufacturerPublished,
            authority: .official,
            confidence: .high,
            publisher: "Kodak",
            title: "KODAK PROFESSIONAL TRI-X 400 Film — Technical Data",
            citation: "Publication F-4017"
        ),
        rules: [
            .formula(FormulaReciprocityRule(
                formula: ReciprocityFormula(
                    formulaFamily: .modifiedSchwarzschild,
                    coefficientSeconds: 2,
                    exponent: 1.3891,
                    noCorrectionThroughSeconds: 0.1,
                    sourceRangeThroughSeconds: 100.0
                ),
                notes: [
                    "App-derived: Tc = 2 × Tm^1.3891, fitted to Kodak's published 1/10/100 sec table rows (no correction through 1/10 sec). Not a Kodak-published formula; the official graph/table model is the default.",
                ]
            )),
        ],
        sourceEvidence: triX400OfficialAnchorEvidence,
        modelBasis: ReciprocityProfileModelBasis(
            sourceModel: .manufacturerTable,
            calculationModel: .guardedFormula
        )
    )

    /// The three official Kodak Tri-X 400 published rows preserved as
    /// source evidence (stop delta + corrected time + development),
    /// shared by the table-only alternate and the app-derived formula
    /// comparison.
    private static let triX400OfficialAnchorEvidence: [ReciprocitySourceEvidenceRow] = [
        triX400AnchorEvidence(metered: 1, stopDelta: 1, corrected: 2, development: "-10% development", note: "1 sec → +1 stop, corrected 2 sec, develop -10%."),
        triX400AnchorEvidence(metered: 10, stopDelta: 2, corrected: 50, development: "-20% development", note: "10 sec → +2 stops, corrected 50 sec, develop -20%."),
        triX400AnchorEvidence(metered: 100, stopDelta: 3, corrected: 1200, development: "-30% development", note: "100 sec → +3 stops, corrected 1200 sec, develop -30%."),
    ]

    private static func triX400AnchorEvidence(
        metered: Double,
        stopDelta: Double,
        corrected: Double,
        development: String,
        note: String
    ) -> ReciprocitySourceEvidenceRow {
        ReciprocitySourceEvidenceRow(
            meteredExposure: .exactSeconds(metered),
            adjustments: [
                .exposure(.stopDelta(StopDeltaAdjustment(stopDelta: stopDelta))),
                .exposure(.correctedTime(CorrectedTimeMapping(meteredSeconds: metered, correctedSeconds: corrected))),
                .development(DevelopmentAdjustment(instruction: development, note: nil)),
            ],
            notes: [note]
        )
    }
}

File: android/core/src/main/kotlin/com/sangwook/ptimer/core/reciprocity/AlternateReciprocityModels.kt
// Copyright © 2026 Sangwook Han
// SPDX-License-Identifier: Apache-2.0

package com.sangwook.ptimer.core.reciprocity

// Faithful port of iOS PTimerCore AlternateReciprocityModels + the
// UnofficialPracticalProfiles registry. Data transcription of the alternate
// models that live OUTSIDE the launch catalog, selectable via the model
// selector. Small builder helpers keep the domain constructors readable.

private fun exact(seconds: Double) =
    MeteredExposureSelector(kind = MeteredExposureSelectorKind.exactSeconds, exactSeconds = seconds)

private fun correctedTimeAdj(metered: Double?, corrected: Double, approximate: Boolean = false) =
    ReciprocityAdjustment(
        kind = ReciprocityAdjustmentKind.exposure,
        exposure = ExposureAdjustment(
            kind = ExposureAdjustmentKind.correctedTime,
            correctedTime = CorrectedTimeMapping(metered, corrected, approximate),
        ),
    )

private fun multiplierAdj(factor: Double) =
    ReciprocityAdjustment(
        kind = ReciprocityAdjustmentKind.exposure,
        exposure = ExposureAdjustment(
            kind = ExposureAdjustmentKind.multiplier,
            multiplier = MultiplierAdjustment(factor),
        ),
    )

private fun stopDeltaAdj(stopDelta: Double) =
    ReciprocityAdjustment(
        kind = ReciprocityAdjustmentKind.exposure,
        exposure = ExposureAdjustment(
            kind = ExposureAdjustmentKind.stopDelta,
            stopDelta = StopDeltaAdjustment(stopDelta),
        ),
    )

private fun developmentAdj(instruction: String) =
    ReciprocityAdjustment(
        kind = ReciprocityAdjustmentKind.development,
        development = DevelopmentAdjustment(instruction = instruction),
    )

private fun tableRule(
    anchors: List<TableAnchor>,
    notes: List<String>,
    noCorrectionThroughSeconds: Double,
    sourceRangeThroughSeconds: Double,
) = ReciprocityRule(
    kind = ReciprocityRuleKind.tableInterpolation,
    tableInterpolation = TableInterpolationReciprocityRule(
        anchors = anchors,
        notes = notes,
        noCorrectionThroughSeconds = noCorrectionThroughSeconds,
        sourceRangeThroughSeconds = sourceRangeThroughSeconds,
    ),
)

private fun formulaRule(formula: ReciprocityFormula, notes: List<String>) = ReciprocityRule(
    kind = ReciprocityRuleKind.formula,
    formula = FormulaReciprocityRule(formula = formula, notes = notes),
)

/**
 * Unofficial practical profiles defined separately from the launch preset
 * catalog (which enforces exactly one official profile per film).
 */
object UnofficialPracticalProfiles {

    fun profile(forFilmID: String): ReciprocityProfile? = when (forFilmID) {
        "kodak-portra-400" -> kodakPortra400UnofficialPractical
        else -> null
    }

    val kodakPortra400UnofficialPractical = ReciprocityProfile(
        id = "kodak-portra-400-unofficial-practical",
        name = "Unofficial practical approximation",
        source = ReciprocitySourceProvenance(
            kind = ReciprocitySourceKind.thirdPartyPublication,
            authority = ReciprocityAuthority.unofficial,
            confidence = ReciprocityConfidence.low,
            publisher = "",
        ),
        rules = listOf(
            formulaRule(
                ReciprocityFormula(
                    formulaFamily = FormulaFamily.modifiedSchwarzschild,
                    exponent = 1.34,
                    noCorrectionThroughSeconds = 0.999_999,
                ),
                notes = emptyList(),
            ),
        ),
        notes = listOf(
            "Unofficial practical approximation. Not a Kodak-published profile.",
            "Formula: Tc = Tm^1.34. Source pending verification.",
        ),
    )
}

/**
 * Registry of alternate reciprocity profiles/models selectable through the
 * model selector via a profile override, never as duplicate top-level film
 * rows.
 */
object AlternateReciprocityModels {

    /** Alternate models for a film stock, in display order. */
    fun alternates(forFilmID: String): List<ReciprocityProfile> = when (forFilmID) {
        "kodak-portra-400" -> listOf(UnofficialPracticalProfiles.kodakPortra400UnofficialPractical)
        "foma-fomapan-100" -> listOf(fomapan100OhzartCommunityTable, fomapan100AppDerivedFormula)
        "kodak-tri-x-400" -> listOf(triX400OfficialTable, triX400AppDerivedFormula)
        "kodak-tmax-100" -> listOf(tmax100AppDerivedFormula)
        "adox-chs-100-ii" -> listOf(chs100AppDerivedFormula)
        else -> emptyList()
    }

    /** Picker display order: primary then alternates, except Tri-X 400. */
    fun modelPickerOrder(primary: ReciprocityProfile, forFilmID: String): List<ReciprocityProfile> {
        val alternates = alternates(forFilmID)
        val officialTable = alternates.firstOrNull { it.id == triX400OfficialTable.id }
        if (forFilmID != "kodak-tri-x-400" || officialTable == null) {
            return listOf(primary) + alternates
        }
        return listOf(officialTable, primary) + alternates.filter { it.id != officialTable.id }
    }

    /** True for explicitly app-derived alternate models. */
    fun isAppDerivedModel(id: String): Boolean =
        id == fomapan100AppDerivedFormula.id ||
            id == triX400AppDerivedFormula.id ||
            id == tmax100AppDerivedFormula.id ||
            id == chs100AppDerivedFormula.id

    /** Resolves an alternate profile by id (used by session restore). */
    fun profile(withID: String): ReciprocityProfile? {
        val all = listOf(
            UnofficialPracticalProfiles.kodakPortra400UnofficialPractical,
            fomapan100OhzartCommunityTable,
            fomapan100AppDerivedFormula,
            triX400OfficialTable,
            triX400AppDerivedFormula,
            tmax100AppDerivedFormula,
            chs100AppDerivedFormula,
        )
        return all.firstOrNull { it.id == withID }
    }

    // MARK: - Fomapan 100 (Ohzart community table + app-derived formula)

    private fun ohzartAnchorEvidence(metered: Double, corrected: Double) = ReciprocitySourceEvidenceRow(
        meteredExposure = exact(metered),
        adjustments = listOf(correctedTimeAdj(metered, corrected)),
    )

    private val ohzartCommunityAnchorEvidence = listOf(
        ohzartAnchorEvidence(1.0, 1.9),
        ohzartAnchorEvidence(2.0, 5.0),
        ohzartAnchorEvidence(4.0, 13.0),
        ohzartAnchorEvidence(8.0, 35.0),
        ohzartAnchorEvidence(15.0, 90.0),
        ohzartAnchorEvidence(30.0, 265.0),
        ohzartAnchorEvidence(60.0, 795.0),
    )

    val fomapan100OhzartCommunityTable = ReciprocityProfile(
        id = "foma-fomapan-100-ohzart-community-table",
        name = "Ohzart community table",
        source = ReciprocitySourceProvenance(
            kind = ReciprocitySourceKind.thirdPartyPublication,
            authority = ReciprocityAuthority.unofficial,
            confidence = ReciprocityConfidence.medium,
            publisher = "Ohzart",
            title = "Reciprocity practical table",
            citation = "https://ohzart1.tistory.com/78",
        ),
        rules = listOf(
            tableRule(
                anchors = listOf(
                    TableAnchor(1.0, 1.9),
                    TableAnchor(2.0, 5.0),
                    TableAnchor(4.0, 13.0),
                    TableAnchor(8.0, 35.0),
                    TableAnchor(15.0, 90.0),
                    TableAnchor(30.0, 265.0),
                    TableAnchor(60.0, 795.0),
                ),
                notes = listOf(
                    "Ohzart community practical table for Fomapan 100, reproduced by log-log interpolation between the published anchors. Practical / community guidance, not FOMA-published data.",
                ),
                noCorrectionThroughSeconds = 0.5,
                sourceRangeThroughSeconds = 60.0,
            ),
        ),
        notes = listOf("Unofficial practical community table (Ohzart). Not FOMA-published data."),
        sourceEvidence = ohzartCommunityAnchorEvidence,
        modelBasis = ReciprocityProfileModelBasis(
            sourceModel = ReciprocitySourceModel.practicalCommunityGuidance,
            calculationModel = ReciprocityCalculationModel.tableLogLogInterpolation,
        ),
        selectorLabel = "Ohzart",
    )

    private fun fomapanAnchorEvidence(metered: Double, multiplier: Double, corrected: Double, note: String) =
        ReciprocitySourceEvidenceRow(
            meteredExposure = exact(metered),
            adjustments = listOf(multiplierAdj(multiplier), correctedTimeAdj(metered, corrected)),
            notes = listOf(note),
        )

    private val fomapanOfficialAnchorEvidence = listOf(
        fomapanAnchorEvidence(1.0, 2.0, 2.0, "1 sec → ×2 (corrected 2 sec)."),
        fomapanAnchorEvidence(10.0, 8.0, 80.0, "10 sec → ×8 (corrected 80 sec)."),
        fomapanAnchorEvidence(100.0, 16.0, 1600.0, "100 sec → ×16 (corrected 1600 sec)."),
    )

    val fomapan100AppDerivedFormula = ReciprocityProfile(
        id = "foma-fomapan-100-app-formula",
        name = "App-derived formula",
        source = ReciprocitySourceProvenance(
            kind = ReciprocitySourceKind.manufacturerPublished,
            authority = ReciprocityAuthority.official,
            confidence = ReciprocityConfidence.high,
            publisher = "FOMA BOHEMIA",
            title = "FOMAPAN 100 CLASSIC — Technical sheet",
            citation = "Foma technical sheet",
        ),
        rules = listOf(
            formulaRule(
                ReciprocityFormula(
                    formulaFamily = FormulaFamily.modifiedSchwarzschild,
                    coefficientSeconds = 2.2457,
                    exponent = 1.4515,
                    noCorrectionThroughSeconds = 0.5,
                    sourceRangeThroughSeconds = 100.0,
                ),
                notes = listOf(
                    "App-derived: Tc = 2.2457 × Tm^1.4515, a free log-log fit through FOMA's published 1/10/100 sec anchors. Not manufacturer-published guidance; the official table model is the default.",
                ),
            ),
        ),
        sourceEvidence = fomapanOfficialAnchorEvidence,
        modelBasis = ReciprocityProfileModelBasis(
            sourceModel = ReciprocitySourceModel.manufacturerTable,
            calculationModel = ReciprocityCalculationModel.guardedFormula,
        ),
    )

    // MARK: - Kodak Tri-X 400 (official table + app-derived formula)

    private fun triX400AnchorEvidence(
        metered: Double,
        stopDelta: Double,
        corrected: Double,
        development: String,
        note: String,
    ) = ReciprocitySourceEvidenceRow(
        meteredExposure = exact(metered),
        adjustments = listOf(
            stopDeltaAdj(stopDelta),
            correctedTimeAdj(metered, corrected),
            developmentAdj(development),
        ),
        notes = listOf(note),
    )

    private val triX400OfficialAnchorEvidence = listOf(
        triX400AnchorEvidence(1.0, 1.0, 2.0, "-10% development", "1 sec → +1 stop, corrected 2 sec, develop -10%."),
        triX400AnchorEvidence(10.0, 2.0, 50.0, "-20% development", "10 sec → +2 stops, corrected 50 sec, develop -20%."),
        triX400AnchorEvidence(100.0, 3.0, 1200.0, "-30% development", "100 sec → +3 stops, corrected 1200 sec, develop -30%."),
    )

    val triX400OfficialTable = ReciprocityProfile(
        id = "kodak-tri-x-official-table",
        name = "Official Kodak table",
        source = ReciprocitySourceProvenance(
            kind = ReciprocitySourceKind.manufacturerPublished,
            authority = ReciprocityAuthority.official,
            confidence = ReciprocityConfidence.high,
            publisher = "Kodak",
            title = "KODAK PROFESSIONAL TRI-X 400 Film — Technical Data",
            citation = "Publication F-4017",
        ),
        rules = listOf(
            tableRule(
                anchors = listOf(
                    TableAnchor(1.0, 2.0),
                    TableAnchor(10.0, 50.0),
                    TableAnchor(100.0, 1200.0),
                ),
                notes = listOf(
                    "Published Kodak E-31 table rows only (1 sec → 2 sec, 10 sec → 50 sec, 100 sec → 1200 sec). No adjustment through 1/10 sec; log-log interpolation between the published anchors. Inputs above 100 sec are flagged beyond the published source range.",
                ),
                noCorrectionThroughSeconds = 0.1,
                sourceRangeThroughSeconds = 100.0,
            ),
        ),
        sourceEvidence = triX400OfficialAnchorEvidence,
        modelBasis = ReciprocityProfileModelBasis(
            sourceModel = ReciprocitySourceModel.manufacturerTable,
            calculationModel = ReciprocityCalculationModel.tableLogLogInterpolation,
        ),
        selectorLabel = "Official table",
    )

    val triX400AppDerivedFormula = ReciprocityProfile(
        id = "kodak-tri-x-app-formula",
        name = "App formula",
        source = ReciprocitySourceProvenance(
            kind = ReciprocitySourceKind.manufacturerPublished,
            authority = ReciprocityAuthority.official,
            confidence = ReciprocityConfidence.high,
            publisher = "Kodak",
            title = "KODAK PROFESSIONAL TRI-X 400 Film — Technical Data",
            citation = "Publication F-4017",
        ),
        rules = listOf(
            formulaRule(
                ReciprocityFormula(
                    formulaFamily = FormulaFamily.modifiedSchwarzschild,
                    coefficientSeconds = 2.0,
                    exponent = 1.3891,
                    noCorrectionThroughSeconds = 0.1,
                    sourceRangeThroughSeconds = 100.0,
                ),
                notes = listOf(
                    "App-derived: Tc = 2 × Tm^1.3891, fitted to Kodak's published 1/10/100 sec table rows (no correction through 1/10 sec). Not a Kodak-published formula; the official graph/table model is the default.",
                ),
            ),
        ),
        sourceEvidence = triX400OfficialAnchorEvidence,
        modelBasis = ReciprocityProfileModelBasis(
            sourceModel = ReciprocitySourceModel.manufacturerGraphTable,
            calculationModel = ReciprocityCalculationModel.guardedFormula,
        ),
    )

    // MARK: - Kodak T-MAX 100 app-derived formula

    private val tmax100OfficialAnchorEvidence = listOf(
        ReciprocitySourceEvidenceRow(
            meteredExposure = exact(1.0),
            adjustments = listOf(
                stopDeltaAdj(1.0 / 3.0),
                correctedTimeAdj(1.0, 1.2599210498948732, approximate = true),
            ),
            notes = listOf("1 sec → +1/3 stop (≈1.26 sec derived time-equivalent)."),
        ),
        ReciprocitySourceEvidenceRow(
            meteredExposure = exact(10.0),
            adjustments = listOf(stopDeltaAdj(0.5), correctedTimeAdj(10.0, 15.0)),
            notes = listOf("10 sec → +1/2 stop, corrected 15 sec."),
        ),
        ReciprocitySourceEvidenceRow(
            meteredExposure = exact(100.0),
            adjustments = listOf(stopDeltaAdj(1.0), correctedTimeAdj(100.0, 200.0)),
            notes = listOf("100 sec → +1 stop, corrected 200 sec."),
        ),
    )

    val tmax100AppDerivedFormula = ReciprocityProfile(
        id = "kodak-tmax-100-app-formula",
        name = "App formula",
        source = ReciprocitySourceProvenance(
            kind = ReciprocitySourceKind.manufacturerPublished,
            authority = ReciprocityAuthority.official,
            confidence = ReciprocityConfidence.high,
            publisher = "Kodak",
            title = "KODAK PROFESSIONAL T-MAX 100 Film — Technical Data",
            citation = "Publication F-4016",
        ),
        rules = listOf(
            formulaRule(
                ReciprocityFormula(
                    formulaFamily = FormulaFamily.modifiedSchwarzschild,
                    coefficientSeconds = 1.2364,
                    exponent = 1.1003,
                    noCorrectionThroughSeconds = 0.1,
                    sourceRangeThroughSeconds = 100.0,
                ),
                notes = listOf(
                    "App-derived: Tc = 1.2364 × Tm^1.1003, fitted to Kodak's published 1/10/100 sec table rows (no correction through 1/10 sec). Not a Kodak-published formula; the official table model is the default.",
                ),
            ),
        ),
        sourceEvidence = tmax100OfficialAnchorEvidence,
        modelBasis = ReciprocityProfileModelBasis(
            sourceModel = ReciprocitySourceModel.manufacturerTable,
            calculationModel = ReciprocityCalculationModel.guardedFormula,
        ),
    )

    // MARK: - ADOX CHS 100 II app-derived formula

    private fun chs100AnchorEvidence(metered: Double, multiplier: Double, corrected: Double, note: String) =
        ReciprocitySourceEvidenceRow(
            meteredExposure = exact(metered),
            adjustments = listOf(multiplierAdj(multiplier), correctedTimeAdj(metered, corrected)),
            notes = listOf(note),
        )

    private val chs100OfficialAnchorEvidence = listOf(
        chs100AnchorEvidence(2.0, 1.5, 3.0, "2 sec → ×1.5 (corrected 3 sec)."),
        chs100AnchorEvidence(4.0, 2.0, 8.0, "4 sec → ×2 (corrected 8 sec)."),
        chs100AnchorEvidence(8.0, 2.5, 20.0, "8 sec → ×2.5 (corrected 20 sec)."),
        chs100AnchorEvidence(15.0, 3.0, 45.0, "15 sec → ×3 (corrected 45 sec)."),
    )

    val chs100AppDerivedFormula = ReciprocityProfile(
        id = "adox-chs-100-ii-app-formula",
        name = "App formula",
        source = ReciprocitySourceProvenance(
            kind = ReciprocitySourceKind.manufacturerPublished,
            authority = ReciprocityAuthority.official,
            confidence = ReciprocityConfidence.high,
            publisher = "ADOX",
            title = "ADOX CHS 100 II S/W Film — Technische Beschreibung, 11. Juli 2024",
            citation = "ADOX CHS 100 II technical sheet (11 July 2024)",
        ),
        rules = listOf(
            formulaRule(
                ReciprocityFormula(
                    formulaFamily = FormulaFamily.modifiedSchwarzschild,
                    coefficientSeconds = 1.2102,
                    exponent = 1.3423,
                    noCorrectionThroughSeconds = 1.0,
                    sourceRangeThroughSeconds = 15.0,
                ),
                notes = listOf(
                    "App-derived: Tc = 1.2102 × Tm^1.3423, fitted to ADOX's published 2/4/8/15 sec table rows (no correction through 1 sec). Not an ADOX-published formula; the official table model is the default.",
                ),
            ),
        ),
        sourceEvidence = chs100OfficialAnchorEvidence,
        modelBasis = ReciprocityProfileModelBasis(
            sourceModel = ReciprocitySourceModel.manufacturerTable,
            calculationModel = ReciprocityCalculationModel.guardedFormula,
        ),
    )
}
